File: challs/sqli.dir/sqli.pl
#! /usr/bin/env perl

use strict;
use warnings;
use DBI;
use Mojolicious;
use Mojo::Log;
use Mojo::URL;
use Mojo::UserAgent;

my $driver = "SQLite";
my $dsn = "DBI:$driver:dbname=/tmp/sqli.db";
my $dbh = DBI->connect($dsn, undef, undef, { AutoCommit => 1 }) or exit;

my $app = Mojolicious->new;
$app = $app->log(Mojo::Log->new(path => '/dev/null'));


sub getUser {
    #
    #    Get account details for user
    #
    my $username = shift;
    my $password = shift;
    my $sth = $dbh->prepare("SELECT username FROM users WHERE username='$username' AND password='$password'");
    $sth->execute;
    $sth->bind_columns(\my($user));
    $sth->fetch;

    return $user;
}

######################   API   ##############################

$app->routes->post('/login' => sub {
    #
    #    Get secret page
    #
    my $c = shift;
    my $username = $c->param('username');
    my $password = $c->param('password');
    
    my $user = eval {getUser($username, $password)};
    if(!$user) {
        return $c->render(
            status => 401,
            template => 'unauthorized',
            text => "Invalid username/password"
        );
    }

    open my $fh, '<', 'secret' or die "error opening secret file";
    my $secret = do { local $/; <$fh> };
        
    return $c->render(
        status => 200,
        text => "You are logged in. The secret is $secret"
    );
});

####################     Main   #############################

sub main {

    my $username = $ARGV[0];
    my $password = $ARGV[1];
    
    my $url = Mojo::URL->new('/login');
    my $ua = Mojo::UserAgent->new();
    $ua->server->app($app);
    my $response = $ua->post($url => form => {username => $username, password => $password});
    print $response->res->content->asset->{'content'};
    $dbh->disconnect();
}

main();
1;

__DATA__

@@ unauthorized.html.ep
% layout 'default';
<h1>Unauthorized</h1>
<p><%= $content %></p>

@@ layouts/default.html.ep
<!DOCTYPE html>
<html>
  <head><title>MyApp</title></head>
  <body><%= content %></body>
</html>

File: challs/sqli.dir/sqli.py
#!/usr/bin/python3

import sqlite3
import sys

from flask import Flask, g, request
from werkzeug.exceptions import InternalServerError, Unauthorized


def get_user(username, password):
    """
        Fetch user in database
    """
    g.cursor.execute(
        "SELECT username FROM users WHERE username='{}' AND password='{}'".format(
            username,
            password
        )
    )
    return g.cursor.fetchone()


def create_app():
    """
        Initialize Flask application
    """
    app = Flask(__name__)

    @app.route('/login', methods=['POST'])
    def get_secret_page():
        """
            Get secret page
        """
        username = request.form['username']
        password = request.form['password']

        try:
            user = get_user(username, password)
            if not user:
                raise Unauthorized('Invalid username/password')
        except sqlite3.OperationalError:
            raise InternalServerError('Something went wrong')

        with open('secret') as _file:
            secret = _file.read().strip()

        return 'You are logged in. The secret is {}'.format(secret)

    return app


APP = create_app()
APP.config['DEBUG'] = True
APP.config['TESTING'] = True


if __name__ == '__main__':

    # Parse params
    if len(sys.argv) != 3:
        print("Missing parameters")
        sys.exit(0)

    username = sys.argv[1]
    password = sys.argv[2]

    if not all((username, password)):
        print("Missing username or password")
        sys.exit(0)

    # Init app and db cursor
    try:
        tester = APP.test_client()
        ctx = APP.test_request_context()
        ctx.push()
        conn = sqlite3.connect('/tmp/sqli.db', isolation_level=None)
        cursor = conn.cursor()
        g.cursor = cursor
    except:
        print('Error while connecting to db.')
        sys.exit(1)

    # Make request
    response = tester.post(
        '/login',
        data=dict(
            username=username,
            password=password
        )
    )

    print(response.get_data().decode('unicode_escape'))
